lib/lightning/oban_manager.ex

```diff
@@ -0,0 +1,203 @@
+defmodule ObanManager do
+  @moduledoc """
+  The Oban Manager
+  """
+  require Logger
+
+  @doc """
+  handles oban events
+  """
+  def handle_event([:oban, :circuit, :open], _measure, meta, _pid),
+    do: Logger.info("Circuit open #{inspect(meta, pretty: true)}")
+
+  def handle_event([:oban, :circuit, :trip], _measure, meta, _pid) do
+    Logger.error("Circuit tripped with #{inspect(meta, pretty: true)}")
+
+    context =
+      Map.take(meta, [:name])
+      |> scrub_context
+
+    Sentry.capture_exception(meta.error,
+      stacktrace: meta.stacktrace,
+      message: meta.message,
+      extra: context,
+      tags: %{type: "oban"}
+    )
+  end
+
+  def handle_event([:oban, :job, :exception], measure, meta, _pid) do
+    Logger.error(~s"""
+    Oban exception:
+    #{inspect(meta.error)}
+    #{Exception.format_stacktrace(meta.stacktrace)}
+    meta:
+      #{Map.drop(meta, [:error, :stacktrace]) |> inspect(pretty: true)}
+    (#{inspect(measure, pretty: true)})
+    """)
+
+    context =
+      meta
+      |> Map.take([:id, :args, :queue, :worker])
+      |> Map.merge(measure)
+      |> scrub_context
+
+    # args = meta.args
+    # queue = meta.queue
+    error = meta.error
+
+    # dead? = meta.attempt >= meta.max_attempts
+    timeout? = Map.get(error, :reason) == :timeout
+
+    if timeout? do
+      Sentry.capture_message("Processor Timeout",
+        level: "warning",
+        message: error,
+        extra: context,
+        tags: %{type: "timeout"}
+      )
+    else
+      Sentry.capture_exception(error,
+        stacktrace: meta.stacktrace,
+        message: error,
+        error: error,
+        extra: context,
+        tags: %{type: "oban"}
+      )
+    end
+
+    # Note: if this is a failed run, we need to ensure that it gets an exit code
+    # and that any subsequent "catch" jobs are initialized. This is done via a
+    # task because—at this point—the best we can do is "try" and we don't want
+    # to tie up the success of this Oban event handler with the success of our
+    # "rescue effort".
+    # if dead? && String.contains?(queue, "runs"),
+    #   do:
+    #     Task.start(__MODULE__, :handle_crashed_run, [
+    #       get_in(args, ["run", "id"]),
+    #       get_in(args, ["state"]),
+    #       error,
+    #       timeout?
+    #     ])
+
+    # handle_crashed_run(
+    #     get_in(args, ["run", "id"]),
+    #     get_in(args, ["state"]),
+    #     error,
+    #     timeout?
+    #   )
+  end
```

Comment:
@elias-ba I'd like there to be a unit test for this function at least. What we discussed was that we weren't going to test that the function is _called_ (we can assume Oban is tested) - what do **do** want to test is that when it's called that it marks the job as finished.

Reply:
Done